### ISSUE TITLE
several cache-related questions

### ISSUE THREAD

**Issue by dreyks** [apotonick/cells#401]
1. It seems that cache does not take into consideration neither template not cell code change. Rails does template digesting to combat this
2. Say I want only a part of cell template cached: common use-case is "cached html with uncached js". What is the best way to achieve this?

Currently I've come up with something like this

```ruby
class MyCell < Cell::ViewModel
  cache :html { 'my_cache_key' }

  def show
    call(:html) + render(:js)
  end

  def html
    render :show
  end
end
```

And then I have `show.erb` which has all the (cached) html, and (uncached) `js.erb` that has all the access to cell instance variables

What d'you think?

**Comment by apotonick** [comment]
I can only reply to 2.) and I think it looks totally fine to me. The point of Cells is to get away from ugly view fragment caching by modelling your view as an object (with multiple states, if you want that).

Only `call` with rails-cells is `html_safe`ed, render really only returns the string.

**Comment by dreyks** [comment]
Ok, got it, thanks

But what about code change? Right now I have to `Rails.cache.clear` after each deploy, which is sub-optimal

**Comment by apotonick** [comment]
Bring it on, sounds like a good feature!

**Comment by dreyks** [comment]
This is not as straightforward to implement as I thought. Rails' `cache` is a view helper, and therefore it knows from which template it was called. Cells' cache on the other hand is called without any knowledge about the template that would render.

Will look into this closer when I have free time

**Comment by apotonick** [comment]
Yeah, but we have a better internal API where you can find the respective state template and hash it, on the class level. No problem!

I'm here if you need help - maybe join Gitter and we chat there at some point.

**Comment by dreyks** [comment]
Yeah, that's what I thought too, but then again, someone can do
```ruby
def show
  render :surprise
end
```
and how will we know that... 

How do I find the chat-room in gitter?

**Comment by apotonick** [comment]
https://gitter.im/trailblazer/chat

**Comment by dreyks** [comment]
#402

**Comment by schorsch** [comment]
I would not dive into this template cache digest thing as it over-complicates  things.

We rather should go for a simpler cache getter/cleaning methods in the core.

f. ex. this is some code i am using to delete the cache for specific cells:

```ruby
class SomeCell

  cache :show do
    "#{model.id}/#{model.updated_at.strftime('%Y%m%d%H%M%S')}"
  end

  cache :public_show do
    "#{model.id}/#{model.updated_at.strftime('%Y%m%d%H%M%S')}"
  end

  # Delete the cache for a single or multiple states (views)
  # If states is empty deletes all defined caches
  # Needed whenever a view is updated, for manual deletion of the cache:
  #
  #   Model.select(:id, :updated_at).each{|i| SomeCell.delete_cache( i, :show)}
  #
  # @param [aModel] obj
  # @param [Array<Symbol>, Symbol] states name of cache partial see defs on top
  def self.delete_cache(obj, states)
    state_keys = states.is_a?(Array) ? states : [states]
    state_keys ||= version_procs.keys
    state_keys.each do |i|
      key = cache_key(i, obj)
      #btw. this does not seem to work maybe bcs my local prefix 'cache' in rails.cache_config is not prepended 
      ::Rails.cache.delete(key)   
    end
  end

  # Get a cache key for a given object and state
  # @param [aModel] obj
  # @param [Array<Symbol>, Symbol] states name of cache partial see defs on top
  # @return [String]
  def self.cache_key(state, obj)
    cell = self.(obj)
    # ??
    state_cache_key(state, version_procs[state].evaluate(cell))
  end
end
```

the specs continue to have a rather ugly lookup:
```ruby
  describe 'cache keys' do
    it 'has keys' do
      obj = FactoryGirl.create(:a_model)
      cell = cell(:some, obj)
      # WTF ??
      key = cell.class.state_cache_key(:show, cell.class.version_procs[:show].evaluate(cell))
      expect(key).to eq ("cells/some/show/#{obj.id}/#{obj.updated_at.strftime('%Y%m%d%H%M%S')}")
    end
  end
```

This could be continued with methods to destroy all caches for a cell, all caches for an object across cells, etc. I am using redis + gem readthis for caching and think about to just use http://redis.io/commands/keys command, which by the help of *-lookups would render all of the above useless.

**Comment by dreyks** [comment]
Yeah, after spending some time fiddling with template digesting I gave up and now use a Capistrano task that clears cell caches for cells whose file were changed between deploys

**Comment by damien-roche** [comment]
Has anything changed? Recently run into this issue myself. I simply added a 'version' to the cell as part of my cache key, and I change that if I update the code. Not ideal, but it works.

**Comment by dreyks** [comment]
yeah, i ended up doing something similar two years ago. i'm not using cells right now though
